### PR TITLE
openblas: add an easy way to compile with DYNAMIC_ARCH=1 enabled.

### PR DESCRIPTION
**Pull request by fcannini**
This is useful in heterogeneous environments where you may be installing the same library for different Intel and/or AMD architectures.

### CODE REVIEW

**Comment by jabl** [comment]
Poking into a (potential?) hornets nest, I'd even argue that DYNAMIC_ARCH=1 should be the default. Not only for heterogeneous environments, but also environments where the build host(s) may not match the execution hosts.

**Comment by chissg** [comment]
Please verify this recipe on macOS: our own non-Spack builds of this package have encountered some problems, especially when building the LAPACK implementation with `DYNAMIC_ARCH=1`.

**Comment by fcannini** [comment]
@jabl @healther Please check bec81e9.

**Comment by davydden** [comment]
@fcannini 

could you please fix indentation:

> var/spack/repos/builtin/packages/openblas/package.py:43: [E501] line too long (81 > 79 characters)
> var/spack/repos/builtin/packages/openblas/package.py:44: [E501] line too long (81 > 79 characters)

**Comment by fcannini** [comment]
Hi everybody.
Do we have a decision on whether to set DYNAMIC_ARCH=1 as default?
IMHO  dealii/dealii#7891 is another reason to do so and avoid similar issues in the future.

@davydden Sure thing. However, the lines shown in the error message will need some rewording if it is decided to change the default `cpu_target`, so I would like to "kill two birds with one stone" and fix all issues at once.

**Comment by fcannini** [comment]
Ping?

File: var/spack/repos/builtin/packages/openblas/package.py [hunk 1]
```diff
@@ -40,8 +40,10 @@ class Openblas(MakefilePackage):
     variant('pic', default=True, description='Build position independent code')
 
     variant('cpu_target', default='auto',
-            description='Set CPU target architecture (leave empty for '
-                        'autodetection; GENERIC, SSE_GENERIC, NEHALEM, ...)')
+        description='Set CPU target architecture (leave empty for autodetection;'
```
Comment (healther) [inline var/spack/repos/builtin/packages/openblas/package.py]: these lines are too long for the flake8 tests

**Comment by aprokop** [comment]
I just noticed that this has not been merged, and wanted to help. But I'm not sure what's going on here. The tests indicate that flake8 check has not passed, however, if I take the patch and do `spack flake8` locally, it passes:
```
=======================================================
flake8: running flake8 code checks on spack.

Modified files:
  var/spack/repos/builtin/packages/openblas/package.py
=======================================================

Flake8 checks were clean.
```

**Comment by alalazo** [comment]
Given the recent support for specific microarchitectures, can't we go for the opposite route:
1. Remove the `cpu_target` variant
2. Add flags and features based on tests for the actual target (see [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#querying-the-target-microarchitecture))

? Would that be possible with `openblas` or people here foresee issues with that?

**Comment by aprokop** [comment]
@alalazo I don't think it solves the same problem. From what I understand, DYNAMIC_ARCH=1 allows to support for more than just one CPU type in a single openblas installation.

**Comment by alalazo** [comment]
> I don't think it solves the same problem. From what I understand, DYNAMIC_ARCH=1 allows to support for more than just one CPU type in a single openblas installation.

I have the same understanding of what `DYNAMIC_ARCH=1` does. What I am wondering is if we have a use case for that in `openblas` now that we can query targets. What would be the meaning of:
```
openblas cpu_target=dynamic target=x86_64
``` 
vs. 
```
openblas cpu_target=dynamic target=knl
``` 
etc. ? Off the top of my head I would say that in general every option that involves architecture specific optimizations should be queried via `target` and set based on the queried property. I didn't think through all the corner cases though, so I was wondering if there was one here. Does that seem reasonable?

**Comment by chissg** [comment]
I would definitely advocate in favor of having DYNAMIC_ARCH=1 be default: it enables a single lowest-common-denominator binary package that supports microarchitecture-specific code when installed on machines with different CPUs.
My one caveat for this is that when enabling the full set of microarchitectures for macOS one runs into a command-line length issue when assembling the library. To ensure successful compilation on macOS one might wish to enable only those architectures known to be found on macs.

**Comment by fcannini** [comment]
@alalazo I think I've understood your point. "Now that we have `target=x86_64` which is a totally generic target, maybe we can use it instead of `cpu_target`." 
Am I correct?
If yes, I see no problem.

About the use case of `DYNAMIC_ARCH=1`, this is very useful in heterogeneous environments, where you may have different micro-architectures, e.g.: a (usually small) cluster that has grown over time, each purchase bringing a different  micro-architecture to the mix. In this case you avoid madness by building just one .so to run them all.

**Comment by sethrj** [comment]
I just hit a similar problem to what you're discussing here, though it is with a build that predated the microarchitecture stuff. On our heterogeneous architecture, the node on which I build OpenBLAS works fine, but other nodes (older presumably) that are used for CI but share the same filesystem fail due to illegal instructions. Will it be sufficient to recompile my toolchain with the generic `target=x86_64`? Or do I need to add an additional variant for OpenBLAS (i.e. `cpu_target=GENERIC`) to realize I want to be able to compile on other systems? :\

**Comment by jabl** [comment]
@sethrj : For the time being you can build the (somewhat poorly named) virtual_machine variant, which will build the DYNAMIC_ARCH=1 version of OpenBLAS which contains specialized kernels for all kinds of CPU's and the correct one selected at runtime.

(You might also need the target=x86_64 or cpu_target=GENERIC in order to get the common code to work everywhere?)

**Comment by sethrj** [comment]
Thanks @jabl . Does anyone (@alalazo?) know of any effort to make this package choose flags based on the Spack-deduced microarchitecture settings? I think it would be worth refactoring the `openblas` package to respect the microarchitecture rather than trying to go its own way, leading to unexpected behavior when downstream codes encounter illegal instructions when running a package that claims to be compiled for the generic x86_64 😄

**Comment by fcannini** [comment]
Hi there.
I just wanted to thank everybody who worked on this issue. The efforts and time spent are much appreciated.

**Pull request closed without merging**